[vm] Add module loading and linking design doc for MonoMove (#19426)
aptos-labs/aptos-core: third_party/move/mono-move/docs/loading_linking.md

# Module Loading and Linking

This document covers how MonoMove loads modules from storage, charges gas for loading, handles cache invalidation on upgrades, and links loaded modules.
The core design decouples loading (I/O, gas, deterministic) from linking (in-memory, unmetered, best-effort optimization), and uses cell-based indirection for cross-module calls following the guard/inline-cache pattern from JIT compilers.

## 1. The Problem

Module loading and gas metering is one of the hardest problems in blockchain VM design.
There is no single solution that addresses all scenarios.
This section describes the core challenges.

### Non-deterministic cache across validators

> The module cache differs across validator nodes.

Reviewer: Is this because of the speculative nature of BlockSTM?

Author: Yes, transactions can executed differently (e.g. conditionals) and in different order

Author: Added this as a sentence

When accessing module `A`, it may be cached on one node but not another.
Because a cache miss is expensive (storage I/O, deserialization, verification), gas must be charged for module loading to ensure deterministic execution costs.

### Upgrades invalidate cached data

Move supports module upgrades.
When a module is upgraded, its size changes, its code changes, and any cached data derived from it (linked function pointers, struct layouts) becomes stale.

> Caches that depends on upgraded module must be invalidated.

Reviewer: Nit: might be more accurate to say something like "All loaded data (caches included) that depends on.."

Author: updated

> When a module is upgraded, its size changes, its code changes, and any cached data derived from it (linked function pointers, struct layouts) becomes stale.

Reviewer 2: In the interest of specifying various constraints on the problem, it may be worth saying here that module upgrades are relatively uncommon (and provide any stats if available, in relation to other kinds of transactions).

Author: Added few sentences about this

> Readers of the upgraded module or caches depending on the upgraded module must be invalidated.

Reviewer 2: > readers of the upgraded module .. must be invalidated

Is is a must? They only need to be invalidated if the speculative execution leads one down a divergent behavior?

Author: This is a realm of behaviour validation - e.g. upgrade that does not affect behaviour does not need to be visible, e.g. module which was republished s part of the package but did not change does not need to invalidate readers, but this is an optimization - otherwise it always should. Added a sentence about this.

> ### "Charge once" semantics

Reviewer 2: What is the "charge once" semantics? Can we write it down explicitly (we have said what we could'nt do).

is it: "we only want to charge once for all the module used by a transaction" (or make it more precise with the definition of use)

Reviewer 2: (this is actually described nicely below, maybe just refer to that section, or move that description here)

Author: changed to 'first-access"

If there is a publish in block B2, it cannot cause cache invalidation yet because some B3 may be executing on B0.
The global cache cannot be mutated during speculative execution.

```
Speculative execution tree:

    B0 (committed)
    ├── B1 (speculative)
    │   └── B3 (speculative, builds on B1)
    └── B2 (speculative, publishes module N)
        └── B4 (speculative, builds on B2)

B2's publish of N cannot invalidate the global cache
because B3 is executing on B0 -> B1 (doesn't see B2).
```

## 2. Invalidation: Push vs. Pull

When a module is upgraded, its upstream users must be invalidated.
There are two approaches.

### Push Invalidation

On upgrade, the writer eagerly finds and invalidates all users.
Walk the dependency graph, null out stale entries, mark readers for re-execution.

*Pros:*
- Cache is always clean.
  Readers never see stale data.
- Invalidation work is proportional to the upgrade event, not spread across reads.

*Cons:*
- Requires maintaining a reverse dependency graph (who depends on B?).
  For transitive dependencies this graph is large and expensive to maintain.
- Layout dependencies make push nearly useless (see below).

> - Push invalidation does not work well under speculation with Zaptos (unless we de-optimize code during specultive upgrade for safety and then re-optimize back).

Reviewer 2: ```suggestion
- Push invalidation does not work well under speculation with Zaptos (unless we de-optimize code during speculative upgrade for safety and then re-optimize back).
```

> - On framework upgrade, the blast radius is enormous (can be mitiagted by special-casing framework, but similar risks exist for populat third-party modules).

Reviewer 2: ```suggestion
- On framework upgrade, the blast radius is enormous (can be mitigated by special-casing framework, but similar risks exist for popular third-party modules).
```

| Upgrade cost           | O(dependents), eager            | O(0) at upgrade, O(readers) lazy      |
| Read cost              | No freshness check              | Version check per access              |
| Layout handling        | Degenerates to pull (general)   | Natural fit (push viable for costs)   |
| Speculative execution  | Requires rollback on abort      | Re-execution re-checks, no rollback   |
| Framework upgrade      | Enormous blast radius           | Demand-driven, proportional to use    |

**Conclusion:** Pull invalidation seems the right choice for general module invalidation.
The layout dependency problem makes general push equivalent to pull in the worst case, while pull is simpler, more natural for speculative execution, and proportional to actual work.
The exception is layout cost validation, where push works well because the reverse index (module to layouts using it) is immutable (see Section 4).
This gives a hybrid: pull for general invalidation, push for layout / monomorphization costs.

## 3. Decoupling Loading from Linking

The core insight of MonoMove's module system is that **loading** and **linking** can be cleanly decoupled into two independent phases.

> ### Loading (Phase 1)

Reviewer: Does this phase materialize anything local to the module?

Author: Yes, I was thinking it can convert to `ModuleIR` (non-lowered ops)

Loading is the I/O phase.
It fetches module bytes from storage (on cache miss) or validates cache entries (on cache hit).
Loading is:

- **Consensus-visible**: tracked by Block-STM in the transaction's read-set.
- **Gas-charged**: every module access incurs a gas cost to ensure deterministic execution costs regardless of cache state.
- **Deterministic**: all validators must agree on how much gas was charged.

The loader is responsible for getting modules into memory and ensuring gas is paid.

### Linking (Phase 2)

> Linking is the optimization phase.

Reviewer 2: Optimization phase makes it sound optional, but we need struct layouts, type information, etc.?

> It takes loaded modules and resolves cross-module references: function pointers, struct layouts, type information.s

Reviewer 2: ```suggestion
It takes loaded modules and resolves cross-module references: function pointers, struct layouts, type information.
```

Linking is:

- **Consensus-invisible**: not tracked by Block-STM.

>   Different validators may produce differently-optimized linked forms.

Reviewer: Is this true though? I understand there maybe some validators may have done linking for a pair of modules while some others might have not, but once the cross-module references get materialized, they should be consistent across all validators, right? In other words, this is nothing more than a "lazy linking optimization", no?

Author: This is because of module cache being different. E.g. if you are on node 1 and nothing is cached, linking cannot do anything because of lazy loading. Eventually, runtime will fill pointers. But on node 2 module cache has module `B`, so some module `A` can be linked to `B` directly (reading from cache) because no need to charge gas and it is in cache. Eventually, states "converge"  - but when we link after loading - things can be different

Author: Clarified the comment

3. INSERT: cache.insert(linked)            // Cache optimized format
```

**Step 1** handles I/O and gas.
On cache hit, gas is charged and the cached executable is returned.
On cache miss, the module is fetched from storage, deserialized, and verified, then gas is charged.

**Step 2** receives cache misses from the loader and has read access to the existing global cache.
It resolves function pointers, computes layouts, and optimizes the executable (if possible).
For dependencies that happen to be in cache, the linker may resolve them eagerly (see Section 7: Optimistic Linking).

**Step 3** inserts the linked executables into the cache for future use.

```
Storage --> [LOADING] --> loaded modules --> [LINKING] --> linked form --> [CACHE]
              │                                 │
              │  gas-charged                    │  gas: not charged
              │  deterministic                  │  deterministic: not required
              │  Block-STM tracked              │  total function 
                                                │
                                       Global cache (opportunistic)
```

> This separation is what makes the system work: gas is fully determined by the loading phase, while linking quality is a best-effort optimization that cannot affect correctness or consensus.

Reviewer 2: "cannot affect correctness" needs a bit more explanation as to why.

Author: This is because linking can use modules from the cache which has non-deterministic state. Added a sentence to clarify.

> It also allows to feature gate loading only while evolvign linking unconditionally.

Reviewer 2: ```suggestion
It also allows to feature gate loading only while evolving linking unconditionally.
```

## 4. Gas Metering

### The Core Principle

Never charge gas for objects derived from modules (layouts, linked forms, monomorphized code) directly.
Instead, use a **proxy** which is cheap to validate and invalidate.

> The module itself is the proxy: charge for the module, and derive everything else from it without additional gas.

Reviewer 2: This means the proxy should account for all the "object derivation costs".

### Charge-Once Semantics

Within a single transaction, each module is charged at most once.
The transaction maintains a per-transaction visited set (the read-set).
On module access:

- Not in read-set: charge gas, add to read-set.
- Already in read-set: skip charging.

This ensures that a loop calling `B::foo` 1000 times pays the loading cost for `B` once.

> ### Layout Costs

Reviewer: Is this an additional charge on top of the module costs? In addition, is this charged once per transaction as well?

Author: Additional charge, but we never charge for modules in this layout. For example, if transaction used `A` and then used layout from `A`, `B`, `C`, (with `A` having max size) its charge will be `A` for module, and `f(A)` for layout.

> Struct layout computation is where gas metering gets subtle.

Reviewer: Actually, a more radical question here is, whether we really still need materialized layouts? Can't we somehow represent it as a tree to be traversed? This way we might not even need to worry about metering it, no?

Author: We still need materialized layouts to know the size - for monomorphization. We do not need this as a special structure, yes. For example, current design for non generic structs embed pointers to field types directly into definition (because they do not change). So we need this wether we traverse or not because we need sizes, offsets, etc. 😭

A layout miss can trigger loading of multiple modules transitively.
Charging for each constituent module via the read-set handles deduplication, but the cost formula for the layout itself needs care.

The proposed approach: charge layout cost via a compressed formula that depends on the modules used to construct it, minimizing cache hit overhead.

```
cost(layout L) = base_cost
               + c0 * len(modules_used(L))

>                + c1 * max(m for m in modules_used(L))

Reviewer: realistically speaking, c0 and c1 are very small numbers (<1, tinay fraction of a module), right?

Author: Yes! We can bucket module length and max, I did not think too much about formula. Most importantly, I we want non-upgradable metric (number of modules) and cheap to charge on cache hit (max of 1 module, but we can also use an average of min-max).

c1 can be different depending on number of modules.

```

This formula has a key property: the set of modules used for a layout is stable across upgrades because struct definitions are immutable in Move.

> Only module *sizes* can change (a module can grow by adding functions).

Reviewer: This is the part I don't understand just yet. If a struct consists of only structs transitively, then its layout is fixed. Why don't we derive a cost from the actual layout size, rather than involving the module size, which by definition involves irrelevant items (e.g. functions) and could change through upgrades?

Author: Because layout load can cause module load in worst case. If you have a struct `A` with field `b::B`, and `b` is not in cache, during monomorphization you need to compute size and hence end up loading `b`.

Reviewer: But with the proposed formula, don't you still have to load all dependency modules to figure out the largest size?

Author: Only on upgrade validation. If we do not use max we need to:
- On cache hit, traverse all modules to charge gas, extend read-set.
- On upgrade, we invalidate readers.

With this we do
- On cache hit, charge gas based on 1 module (cached), extend read-set with 1 module (for validation only).
- On upgrade, find layouts which contain this module. Check if any of these intersect with read-set.

Basically, we can make cache hit cheaper by making validation recompute a few more things. Unlike transitive closure, this is better because layouts do not change so modules used to construct it are stable.

Reviewer: I see, so the `max` there really is to make it possible to update the cost without collecting the whole transitive closure of the struct.

  This is acceptable because framework upgrades are rare.

TBD: consider interaction with Zaptos.

## 5. Edge Tracking and Metering

An alternative to per-module tracking is tracking and metering *edges* (dependency relationships) rather than individual modules.

### The Idea

Instead of asking "which modules did this transaction load?", ask "which dependency edges did this transaction traverse?".
An edge is a pair `(caller_module, callee_module)` representing a cross-module call or type reference.

### Motivation

If each caller-callee pair has its own inline cache (e.g., each `CallExternal` site remembers its target), then the question "has `B` been charged?" becomes local per-edge.
There is no need for a shared per-transaction set check on every call.
The information is embedded in the edge itself.

### Pros

- Easier to track: per-edge state eliminates shared set lookups.
- Push invalidation may be easier: upgrade to `B` means "find all readers of edges pointing to `B`".
  Each read-set entry is an edge, so the scan is direct.
- Natural fit for inline caches: the edge *is* the cache entry.

### Cons

- Read-set size increases significantly (edges >> modules).

> - Overcharging of popular libraries: every caller of `0x23::math` pays an edge cost, even though from an I/O perspective the module is loaded once.

Reviewer 2: when you say "every caller", this is once per transaction, the edge cost for `module_A` linking to `0x23::math` is paid once (and not once per call site)?

if this is the case, why is it overcharging? each module pays for using a different module, once.
if this is not the case, then we are not asking "which dependency edges did this transaction traverse", but "how many times were each of those dependency edges traversed"?

Author: I meant if you have module A and modules B, C, D, link to it and use these edges, you would charge for each edge incoming into A - so charging it effectively 3 times and not 1. In general, number of edges can be quadratic so I think this is not a good approach anyway 😄

| Read-set size          | Small (unique modules)           | Large (unique edges)                  |
| Charge-once check      | Shared set lookup                | Local per-edge (inline cache)         |
| Push invalidation      | Reverse dependency graph needed  | Scan edges pointing to upgraded module|
| Overcharging risk      | Lower                            | Higher for hot libraries              |
| Block-STM fit          | Natural (modules are the unit)   | Needs mapping edges back to modules   |

**Status:** Node tracking is preferred for now due to simplicity and natural fit with Block-STM's read-set mechanism.

## 6. Cell-Based Linking

Cell-based linking is the core mechanism for cross-module function calls.
It follows the **guard/inline-cache optimization pattern** from JIT compilers: embed a pointer to the target, guard it with a version check, and fall back to slow resolution on mismatch.

### Cell Structure

For every cross-module call target `B::foo`, there is a cell that belongs to module `B`:

```rust
struct FunctionCell {
    /// Fast path: pointer to resolved function + version tag or hash.
    /// Null means "not yet resolved" or "invalidated".

>     base: AtomicPtr<(Function, Version)>,

Reviewer 2: Should version just be part of `Function`?

    /// Speculative overlay: maps speculative block + txn index version to function.
    /// Empty during committed execution. Used by Zaptos for speculative blocks that publish modules.

>     overlay: SmallVec<(Version, *const Function)>,

Reviewer 2: could you explain this part in a bit more detail?

Reviewer 2: (ah, I think this is explained later)
